File: Python/example_25/DTB_model.py
import numpy as np

############################
### Case study 25
############################
def DTB_model(x):
    # ####################################################################### #
    # DTB_model Depth to bedrock landscape evoluton model of Gomez et al.     #
    # ####################################################################### #

    if not hasattr(DTB_model, "initialized"):                      # Store local variables in memory
        # Load input data
        data = np.loadtxt('Input_data.txt')
        # Python - reshape changed as MATLAB counts vertical, Python horizontal!
        DTB_model.Zx = np.reshape(data[1:, 0], (int(data[0, 1]), int(data[0, 0]))).T    # slope gradient x-direction
        # Python - reshape changed as MATLAB counts vertical, Python horizontal!
        DTB_model.Zy = np.reshape(data[1:, 1], (int(data[0, 1]), int(data[0, 0]))).T    # slope gradient y-direction
        # Python - reshape changed as MATLAB counts vertical, Python horizontal!
        Ld = np.reshape(data[1:, 2], (int(data[0, 1]), int(data[0, 0]))).T              # drainage distance
        DTB_model.Ld_n = (Ld - np.min(Ld)) / (np.max(Ld) - np.min(Ld))                  # normalized drainage distance
        DTB_model.Ld = Ld
        DTB_model.nabla_z = DTB_model.Zx ** 2 + DTB_model.Zy ** 2                       # gradient norm
        DTBdata = np.loadtxt('DTB_data.txt')
        id = DTBdata[:, 1].astype(int)                                                  # define training points
        # Turn these indices into valid Python rows and columns
        nr, nc = DTB_model.Ld.shape
        DTB_model.r = np.zeros_like(id)                                                 # Store row indices
        DTB_model.c = np.zeros_like(id)                                                 # Store column indices
        # Loop over the linear indices
        for i in range(len(id)):
            # Column index (0-based in Python)
            DTB_model.c[i] = np.ceil(id[i] / nr) - (1)                                  # Python's 0-based indexing
            # Row index (0-based in Python)
            DTB_model.r[i] = id[i] % nr - (1)                                           # Python 0-based indexing
        DTB_model.initialized = True                                                    # Flag to indicate that initialization is comple

    Lambda = np.exp(-x[1] * (1 - DTB_model.Ld_n) ** x[3])                               # bedrock valley shape term
    Psi = 1 - (DTB_model.nabla_z / x[2] ** 2)                                           # threshold angle of mass movement
    H = (Psi / Lambda) * np.sqrt(x[0] * DTB_model.Ld ** 2)                              # Modeled depth to bedrock
    H = H[DTB_model.r,DTB_model.c]                                                      # Isolate the measurement data points of id
    H = H.flatten()                                                                     # Flatten the matrix to a 1D array

    return H

File: MATLAB/example_25/DTB_model.m
function H = DTB_model(x)
% <><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><> %
% <><><><><>  DTB_model - Simulation of the depth to bedrock  <><><><><> %
% <><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><> %
%                                                 Version:   29.Jan.2016 %
%                                                    (c) Guilherme Gomes %
%                                              guilhermejcg@yahoo.com.br %
%                                            Modified by Jasper A. Vrugt %
% This program simulates the bedrock depth                               %
%                                                                        %
% SYNOPSIS                                                               %
%                                                                        %
% function H = DTB_model(x)                                              %
%                                                                        %
% Input:  x       vector of parameters (Phi, lambda_1, Sc, lambda_2)     %
% Output: H       vector of simulated bedrock depths                     %
%                                                                        %
% <><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><> %

% Retain variables in local memory
persistent Ld Zx Zy Ld_n nabla_z id

if isempty(Ld)
    data = load('Input_data.txt');                      % load input data
    Zx = reshape(data(2:end,1),data(1,1),data(1,2));    % slope gradient x-direction
    Zy = reshape(data(2:end,2),data(1,1),data(1,2));    % slope gradient y-direction
    Ld = reshape(data(2:end,3),data(1,1),data(1,2));    % drainage distance
    Ld_n = (Ld - min(Ld(:)))./(max(Ld(:))-min(Ld(:)));  % normalized drainage distance
    max(Ld(:)), min(Ld(:))
    Ld(1,:)
    size(Ld)
    nabla_z = Zx.^2 + Zy.^2;                            % gradient norm
    DTBdata = load('DTB_data.txt'); id = DTBdata(:,2);  % define calibration points
    % NOTE: Can use 6th variable "plugin" in example_25 
    % instead to port all this stuff as second variable 
    % to this DTB_model function
end

Lambda = exp( - x(2) * (1 - Ld_n).^x(4) );              % bedrock valley shape term
Psi = 1 - ( nabla_z ./ x(3)^2 );                        % threshold angle of mass movement
H = ( Psi ./ Lambda ) .* sqrt ( x(1) * Ld.^2 );         % Modeled depth to bedrock
H = H(id);                                              % return simulated data

end